### ISSUE TITLE
第6回ミーティング議事録

### ISSUE THREAD

**Issue by kou-tanaka-1226** [Keiso-Ken/deep-learning-from-scratch-exercise#16]
# 第6回計装研 数理グループ ミーティング  
  
## 日時

2020/10/22
19:00～21:30

## 場所

オンライン(Zoom)

## 参加者

* KiQ Robotics株式会社  
  * 上原さん  
  * 森田さん
* 北九州市立大学  
  * 永原先生
* 三菱ケミカル株式会社
  * 城間さん
  * 藤田さん
* 株式会社C&Gシステムズ  
  * 手嶋  
  * 田中(議事)

# 輪読

[参考書](https://www.amazon.co.jp/%E3%82%BC%E3%83%AD%E3%81%8B%E3%82%89%E4%BD%9C%E3%82%8BDeep-Learning-%E2%80%95Python%E3%81%A7%E5%AD%A6%E3%81%B6%E3%83%87%E3%82%A3%E3%83%BC%E3%83%97%E3%83%A9%E3%83%BC%E3%83%8B%E3%83%B3%E3%82%B0%E3%81%AE%E7%90%86%E8%AB%96%E3%81%A8%E5%AE%9F%E8%A3%85-%E6%96%8E%E8%97%A4-%E5%BA%B7%E6%AF%85/dp/4873117585)

## 第6章 学習に関するテクニック(担当：森田さん)

* 6.1 パラメータの更新
  * 6.1.1 冒険家の話
  * 6.1.2 SGD
  * 6.1.3 SGDの欠点
  * 6.1.4 Momentum
  * 6.1.5 AdaGrad
  * 6.1.6 Adam
  * 6.1.7 どの更新手法を用いるか？
  * 6.1.8 MNISTデータセットによる更新手法の比較
* 6.2 重みの初期値
  * 6.2.1 重みの初期値を0にする？
  * 6.2.2 隠れ層のアクティベーション分布
  * 6.2.3 ReLUの場合の重みの初期値
  * 6.2.4 MNISTデータセットによる重みの初期値の比較
* 6.3 Batch Normalization
  * 6.3.1 Batch Normalizationのアルゴリズム
  * 6.3.2 Batch Normalizationの評価
* 6.4 正則化
  * 6.4.1 過学習
  * 6.4.2 Weight Decay
  * 6.4.3 Dropout
* 6.5 ハイパーパラメータの検証
  * 6.5.1 検証データ
  * 6.5.2 ハイパーパラメータの最適化
  * 6.5.3 ハイパーパラメータの最適化の実装
* 6.6 まとめ

## 前回の内容についてのコメント

* テンソル積ノードの逆伝播の一般則について（田中）
  * 先週，Affineノードに関して話しているときにテンソル積ノードではどうなるかという話が出たため，考察した。（田中）
  * [資料保管場所] https://drive.google.com/drive/folders/105yJV9qhwZmHHr0Z0ZjcTHR0zkfXh2zX
  * 物理の世界では，テンソルは上付き下付きがあり，和をとるときは上と下に同じ文字があるときで，資料にあるように，下に同じ添え字が2つ現れることは許されなかった気がするが。。。（永原先生）
    * 相対論（リーマン幾何学）では確かにその通りで，このような表記は許されない。しかし，今回のテンソル積ノードを考察する上で，添え字の上付き下付きの区別に意味はないため，添え字はすべて下付きとした。今回の範囲内ではこのようにしても特に問題はない。（田中）
  * 上付き下付きはどういう意味であったか？（永原先生）
    * 相対論（リーマン幾何学）では，計量テンソル<img src="https://latex.codecogs.com/gif.latex?g_{\mu\nu}"/>, <img src="https://latex.codecogs.com/gif.latex?g^{\mu\nu}"/>というものをテンソルにかける（上げ下げする対象の添え字について和をとる）ことによって添え字が上げ下げされる。（田中）（補足：例　<img src="https://latex.codecogs.com/gif.latex?A_{i}=g_{i\nu}A^{\nu}"/>，<img src="https://latex.codecogs.com/gif.latex?T_{i}^{\:&space;j}=g^{j\nu}T_{i&space;\nu}"/>など。空間の距離の定義が計量テンソルを用いて行われている場合には，この上付き下付き記法が便利であり，相対論ではこの記法が基本になっている。）
    * （座標変換の仕方の違いにより）上付き添え字のみのテンソルは反変テンソル，下付き添え字のみのテンソルは共変テンソルと呼ばれる。（田中）（補足：両方持つものは混合テンソルと呼ばれる。）
  * 画像の世界ではテンソルがよく登場する（永原先生）
    * 画像の世界では上付き下付き添え字は登場する？（田中）
      * しないと思う（永原先生） 

* Softmax with Lossレイヤの逆伝播（N変数）の導出（田中）
  * 永原先生に先週2変数で導出していただいたので，自分でもN変数で導出してみた。（田中）
  * [資料保管場所] https://drive.google.com/drive/folders/105yJV9qhwZmHHr0Z0ZjcTHR0zkfXh2zX
  * 計算してみてこの結果を導出することは容易いが，最初にこれ（Cross Entropyコスト関数）を考えついた人は偉大である。（永原先生）

## 本章に対してのコメント

* 確率的勾配降下法
  * 本章において「確率的」勾配降下法と記述しているが，全訓練データを使用して<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;L}{\partial&space;W}"/>を真面目に計算するときは単純に勾配降下法という。厳密に計算すると時間がかかるので，入力データからいくつかデータを選択して，<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;L}{\partial&space;W}"/>を近似的に計算するのが確率的勾配降下法である。（永原先生）
    * 「確率的」ではないときの正式名称はなんと言うのか？（手嶋）
      * 最急降下法，あるいは勾配降下法（永原先生）

* Momentum
  * Momentumは，力学の慣性の法則をパラメータの更新に適用した形になっている。パラメータ更新において，前回のパラメータ更新時の変化の「勢い」が今回のパラメータ変化にある比率<img src="https://latex.codecogs.com/gif.latex?\alpha"/>で引き継がれる。Momentum（運動量）よりInertial（慣性）という名前の方が適切ではないかと思う。（永原先生）
    * ソースコードで，速度に乗じられた係数はMomentumという名前になっているが，Momentumというより摩擦係数（速度に比例する抵抗の係数）を1から引いたような感じに見えるので，Momentumという名前は適切なのだろうか（田中）

  * パラメータの更新の経路が，制御工学的に自然である。大きな撃力が働いたような，急激に変化する経路にならない。  
  パラメータ更新の式
  <img src="https://latex.codecogs.com/gif.latex?\boldsymbol{v}\leftarrow\alpha\boldsymbol{v}-\eta\frac{\partial&space;L}{\partial&space;\boldsymbol{W}}"/>
  および
  <img src="https://latex.codecogs.com/gif.latex?\boldsymbol{W}\leftarrow\boldsymbol{W+v}"/>
  は，積分系を通すようなイメージ。<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;L}{\partial&space;\boldsymbol{W}}"/>を重みをつけて積分するようなものになっている。この辺は機械系の話で，力学と関係がある（永原先生）

    * 「焼きなまし法」というものも物理学由来の方法（森田さん）
    * 機械系・物理系の人が真面目に考えれば，もっとうまくやる方法を思いつくことができるかもしれない。PID制御の応用など考えられる（永原先生）

  * Momentumは制御工学で言うところのローパスフィルターのようなもの。（永原先生）

* AdaGrad

  * 式(6.6)はいったいどういう演算をしている？（田中）
    * これは不親切な（というより一貫性がない）書き方。上では行列要素ごとの積が◎で書かれてるが，下ではそうなっていない。実際は下でも行列要素ごとの積をとる式になっている。（永原先生）  
    （補足：正しく書くなら以下のようになる。）
    <img src="https://latex.codecogs.com/gif.latex?\boldsymbol{h}\leftarrow\boldsymbol{h}+\frac{\partial&space;L}{\partial&space;\boldsymbol{W}}\odot\frac{\partial&space;L}{\partial&space;\boldsymbol{W}}"/>
    として，
    <img src="https://latex.codecogs.com/gif.latex?\boldsymbol{h'}"/>を<img src="https://latex.codecogs.com/gif.latex?[\boldsymbol{h'}]_{ij}=\frac{1}{\sqrt{[\boldsymbol{h}]_{ij}}}"/>という要素を持つ行列として
    <img src="https://latex.codecogs.com/gif.latex?\boldsymbol{W}\leftarrow\boldsymbol{W}-\eta&space;\boldsymbol{h'}\odot\&space;\frac{\partial&space;L}{\partial&space;\boldsymbol{W}}"/>

  * ちなみにMATLABでは，行列で割ると表現は，インバース（逆行列）をかけるということを表すので注意。（永原先生）

  * 割り算の分母に(ゼロ割を防ぐ目的で)微小量を足すのは一般的？（手嶋）
    * よくやる方法ではある（永原先生）
    * 変数の型が明示的でないPythonにおいて，微小量を決め打ちして大丈夫なのか？適切な微小量というのは変数の型に依存すると思うが。。。（手嶋）
    * これは一種のハイパーパラメータになる。ただ，わりと適当に十分小さい数を設定することが多い。（永原先生）
    * 今はやっている半精度とか，もっと極端に0 or 1とかの変数を使うことも最近あるので，このようなハードコーディングは危険な気がする。（手嶋）
    * 最終的には丸められるので，適当でもよいという考え方でこのようにしている。コスト関数に<img src="https://latex.codecogs.com/gif.latex?L_0"/>ノルムを使ったりする場合は，パラメータに敏感なため問題が生じることがあるが，今回のようにコスト関数がパラメータについて微分連続な場合にはそんなに影響はなく，問題にならないと思われる。（永原先生）

* Adam
  * 一般的な最適化にはLevenberg-Marquardt法などが使われることがあるが，機械学習の話ではあまり聞かない。（森田さん）

    * Levenberg-Marquardt法ではないが，<img src="https://latex.codecogs.com/gif.latex?\sin(x)"/>を表現する2層のニューラルネットワーク（活性化関数は中間層：シグモイド，出力層：恒等関数）をあえてNewton法で求めようとしたことがあるが，あまりうまくはいっていない。初期値依存性が大きい。（田中）

      * Newton法で最適化を行うと，Hessian（2階偏微分行列）を求めないといけないので時間がかかるのでは。なぜNewton法？（森田さん）
      * 自身が慣れ親しんでいる方法のため，Newton法でやってみた。確かに時間はかかる。そのうえ結果も芳しくない。（田中）
      * Newton法は，二次収束のため，初期値がある程度解に近くないとうまく収束しない。（永原先生）

* 重みの初期値
  * 「He」は「へー」と読む。人名。（永原先生）

* 正則化(Weight decay)
  * <img src="https://latex.codecogs.com/gif.latex?L_\infty"/>ノルムは成分の中の絶対値の最大値になる。<img src="https://latex.codecogs.com/gif.latex?L_0"/>ノルムは成分の中の0でない成分の個数になる。（永原先生）
    * <img src="https://latex.codecogs.com/gif.latex?L_0"/>ノルムは使うメリットがある？（森田さん）
      * パラメータ数が多いときに役に立つ（永原先生）  

* 正則化(Dropout)
  * Weight decayで対応できなくなるというのは，<img src="https://latex.codecogs.com/gif.latex?L_2"/>ノルムをWeight decayで使用している場合，逆行列の計算が必要になってくるので，計算量が多くなってくると対応しきれない。数値計算的な難しさが出てくる。<img src="https://latex.codecogs.com/gif.latex?L_0"/>ノルムを用いると，さらに大変になる。（永原先生）

  * Dropoutで，無効にしたニューロンからの次の層への出力はどうなっている？ある層から次の層への出力は形が決まっているので，その形（要素数）を変えてはまずいのでは？
    * Dropoutは，具体的には出力ベクトルに出力ベクトルと同じ形のマスクベクトル（出力するニューロンに対応する要素は1，無効にしたニューロンに対応する要素は0）をかける（要素ごとの積）ことによって，無効にしたニューロンの出力に対応する要素を0にしている。よって，出力ベクトルの形は変わっていない。（森田さん）

## 本章で学んだこと(参考書より抜粋)

* パラメータの更新方法には，SGDの他に，有名なものとして，MomentumやAdaGrad，Adamなどの手法がある。
* 重みの初期値の与え方は，正しい学習を行う上で非常に重要である。

* 重みの初期値として，「Xavierの初期値」や「Heの初期値」などが有効である。

* Batch Normalizationを用いることで，学習を早く進めることができ，また，初期値に対してロバストになる。

* 過学習を抑制するための正則化の技術として，Weight decayやDropoutがある。

* ハイパーパラメータの探索は，良い値が存在する範囲を徐々に絞りながら進めるのが効率の良い方法である。

# その他

* Githubについて
  * Githubにソースコードを上げる際に，本で参照しているソースコード（たとえばMNISTデータセット読み込みとか）は，本の著者のGithubリポジトリーから持ってきて大丈夫なのか？（手嶋）
    * MITライセンスなので，ライセンス条項を記載していれば問題ない。（森田さん）
    * Forkするのが一番早いのでForkする（ミーティング内でFork実施完了）。（森田さん）
      * Forkすると，おおもとのリポジトリが更新されても，Forkした側は更新されないので注意。Fork後に更新したいなら手動更新になる。（森田さん）

* 今後の方針
  * 次回でこの本の内容（の皆で読む部分）は終了する。今後の活動はどうする？（永原先生）
    * 続編（自然言語処理編）に進む方向でメンバの意見が一致
    * [続編の書籍]https://www.amazon.co.jp/%E3%82%BC%E3%83%AD%E3%81%8B%E3%82%89%E4%BD%9C%E3%82%8BDeep-Learning-%E2%80%95%E8%87%AA%E7%84%B6%E8%A8%80%E8%AA%9E%E5%87%A6%E7%90%86%E7%B7%A8-%E6%96%8E%E8%97%A4-%E5%BA%B7%E6%AF%85/dp/4873118360/ref=pd_bxgy_img_2/356-3423631-0422010?_encoding=UTF8&pd_rd_i=4873118360&pd_rd_r=7d6585d0-1f69-4772-a5b7-1a939d8aa3aa&pd_rd_w=DOPn3&pd_rd_wg=BId6e&pf_rd_p=e64b0a81-ca1b-4802-bd2c-a4b65bccc76e&pf_rd_r=96GXCJ28GBZDMP2YE5KW&psc=1&refRID=96GXCJ28GBZDMP2YE5KW

* 新メンバー候補の村上さんが夜の参加は厳しい件について
  * スプレッドシートに一応村上さんの名前も記載して，参加可能日程アンケートに記入してもらう方向で（上原さん）
  * 個人的には，休日はちょっと厳しいが，平日早朝なら大丈夫（永原先生）

# 次回

* 7章 畳み込みニューラルネットワーク (担当：城間さん，藤田さん)

* 日程はGitHubのissueにて決定(2020/11/05～2020/11/11で開催)
